### PR TITLE
feat: upgrade signed FPs msgs

### PR DESCRIPTION
**Pull request by RafilxTenfen**
- Load signed `MsgCreateFinalityProvider` from json string
- Add finality providers in upgrade
- Check finality providers in e2e upgrade signet

### CODE REVIEW

File: app/keepers/keepers.go
Comment (RafilxTenfen) [inline app/keepers/keepers.go:162]: I needed this for the upgrade to parse the signed msgs, the option was to add it here or at the [CreateUpgradeHandler](https://github.com/babylonlabs-io/babylon/blob/5f8af8ced17d24f3f0c6172293cd37fb3d055807/app/upgrades/types.go#L29)

File: app/upgrades/signetlaunch/upgrades.go
Comment (RafilxTenfen) [inline app/upgrades/signetlaunch/upgrades.go:138]: This is not strictly necessary as all the finality providers would have the same order when we compile, but I still kept it...

File: app/upgrades/signetlaunch/data_signed_fps.go
Comment (SebastianElvis) [inline app/upgrades/signetlaunch/data_signed_fps.go:3]: Is it a hard requirement that the string needs to contain the full tx rather
than just the message part?

File: app/upgrades/signetlaunch/upgrades.go
Comment (SebastianElvis) [inline app/upgrades/signetlaunch/upgrades.go:113]: Hmm if we only have the message part then we don't need to unmarshal things into `any`

File: x/btcstaking/keeper/finality_providers.go
Comment (SebastianElvis) [inline x/btcstaking/keeper/finality_providers.go:17]: Good refactoring! Can `SetFinalityProvider` be made private then?

File: app/upgrades/signetlaunch/upgrades.go
Comment (gitferry) [inline app/upgrades/signetlaunch/upgrades.go:107]: Could the json string load from a file?

File: app/upgrades/signetlaunch/upgrades.go
Comment (gitferry) [inline app/upgrades/signetlaunch/upgrades.go:154]: why? we should explain here

File: app/upgrades/signetlaunch/upgrades.go
Comment (gitferry) [inline app/upgrades/signetlaunch/upgrades.go:176]: Should we ensure the number of signers is 1?

File: app/upgrades/signetlaunch/upgrades.go
Comment (gitferry) [inline app/upgrades/signetlaunch/upgrades.go:143]: should we also check the commission rate to be not less than the min value defined in params?

File: app/upgrades/signetlaunch/data_signed_fps.go
Comment (KonradStaniec) [reply app/upgrades/signetlaunch/data_signed_fps.go:3]: I think I will attach myself to this comments as this is probably closest to my concerns.

I think the question here becomes where do we validates `bbn1gwecky0m842kvjg7mcvt9z330rz96948aplqlm` is if fact the signer of the transactions:

1. We can do it off-chain, in some process /service and here include only the messages ?
2. W can do it on-chain, so gather transactions to some database and during upgrade perform whole transaction validation including transaction signaure. To do that we would need to move all of those transaction through message router as transactions signature validation is performed in `anteHandler`. Then we would ignore the ones that fail the sig validation (though this seems like bad ux ?)

File: app/upgrades/signetlaunch/upgrades.go
Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/upgrades.go:107]: Nop :/ otherwise, it doesn't compile together with the binary, so you would need to also send this json file to who wants to upgrade and put it aside the binary when the upgrade is happening

File: app/upgrades/signetlaunch/upgrades.go
Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/upgrades.go:168]: I will add a check

File: app/upgrades/signetlaunch/upgrades.go
Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/upgrades.go:154]: it is what we expect, for each signer to only submit one message per transaction 
this is the format we are going to expect the finality providers submit to us

File: app/upgrades/signetlaunch/upgrades.go
Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/upgrades.go:176]: agree

File: app/upgrades/signetlaunch/data_signed_fps.go
Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/data_signed_fps.go:3]: yes, we can do it off-chain as option 1. 
or in option 2. I am validating during the upgrade https://github.com/babylonlabs-io/babylon/blob/5215030b9b8e4d51fb63a633cc531c7e4f445224/app/upgrades/signetlaunch/upgrades.go#L177-L178

File: x/btcstaking/keeper/finality_providers.go
Comment (RafilxTenfen) [reply x/btcstaking/keeper/finality_providers.go:17]: done at https://github.com/babylonlabs-io/babylon/pull/28/commits/1b6e6bafea91db6262dc7b40ae0875ad069023ad

File: app/upgrades/signetlaunch/upgrades.go
Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/upgrades.go:143]: The commission rate is already check at https://github.com/babylonlabs-io/babylon/blob/1b6e6bafea91db6262dc7b40ae0875ad069023ad/x/btcstaking/keeper/finality_providers.go#L23

File: app/upgrades/signetlaunch/upgrades.go
Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/upgrades.go]: done at https://github.com/babylonlabs-io/babylon/pull/28/commits/2bb47c4f305f489af2065402ba56576a1ced60cd

File: app/upgrades/signetlaunch/upgrades.go
Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/upgrades.go:173]: https://github.com/babylonlabs-io/babylon/pull/28/commits/2bb47c4f305f489af2065402ba56576a1ced60cd

File: app/upgrades/signetlaunch/data_signed_fps.go
Comment (SebastianElvis) [reply app/upgrades/signetlaunch/data_signed_fps.go:3]: Do we really need to validate the signature of the tx signed by the signer?

It's ourselves who determine `SignedFPsStr` and we will need to gather data and verify them offline anyway. Only after we are confident `SignedFPsStr` is correct we will perform this software upgrade.

So it seems that inserting messages directly is already secure?

Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/data_signed_fps.go]: Updated at https://github.com/babylonlabs-io/babylon/pull/28/commits/79a07120715aa0096c7ddf9c2537721f5627eea5

One thing to note is, that I would still expect finality providers to give us the entire signed message during the registry and we convert it to the Finality provider information as needed 

Mostly as a guarantee that the fp owns that key and is confirming that he wants to be a FP for that BTC pk, now that the PoP only covers BTC owns address :sweat_smile:

**Review by gitferry — APPROVED** [review]
Code-wise lgtm!

File: app/upgrades/signetlaunch/data_signed_fps.go
Comment (KonradStaniec) [reply app/upgrades/signetlaunch/data_signed_fps.go]: what misled me, is that the line you linked before ie
```
cdc.InterfaceRegistry().SigningContext().AddressCodec().BytesToString(signers[0])
```
does not really verify signature. To verify signature easiest way would be copy code from ante handler:
https://github.com/cosmos/cosmos-sdk/blob/main/x/auth/ante/sigverify.go#L302

> One thing to note is, that I would still expect finality providers to give us the entire signed message during the registry and we convert it to the Finality provider information as needed
> 
> Mostly as a guarantee that the fp owns that key and is confirming that he wants to be a FP for that BTC pk, now that the PoP only covers BTC owns babylon address

totally agree on that we need to verify signature for provider Babylon address. 

One things that worries me, if we miss something and add transactions with invalid signature to this string, and now we won't validate it on node level. It seems risky.  Maybe, as this whole string is part of our binary, we can have unit test that goes over whole thing and verify signatures ?

Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/data_signed_fps.go]: Would You prefer to have the full signed message in the string and validate the signature during the upgrade? 
Or validate it offline and simplify the upgrade with just the Finality Provider information? 

About the `sigverify`: 
this improvement in `sigverify` is still unreleased https://github.com/cosmos/cosmos-sdk/blob/0fe3115dcf2e42dac29364a70cca71d52df2a443/x/auth/CHANGELOG.md#L38

The one we currently have loads the account from the AccountKeeper 
and this is not good for us, since the account was not created yet by not submitting any transaction by the finality provider babylon key holder before

image

Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/data_signed_fps.go:3]: As discussed with @KonradStaniec offline, the string with finality providers will contain the entire transaction with signature, but the signatures validity, PoP and other checks will not be verified during the upgrade, only by a unit test `TestValidateSignatureSignedFPsFromData` that should make sure everything is checked for the upgrade to happen on that front

So, during the upgrade we would still need to parse the msg to `Any` to be able to cast it to the `MsgCreateFinalityProvider` after, but without validating other things

cc @SebastianElvis

Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/data_signed_fps.go:3]: I will wait for you guys approval on this before merge

File: app/upgrades/signetlaunch/data_signed_fps_test.go
Comment (KonradStaniec) [inline app/upgrades/signetlaunch/data_signed_fps_test.go:22]: hmm we could probably also add validation if the string do not containan duplicates

File: app/upgrades/signetlaunch/data_signed_fps_test.go
Comment (KonradStaniec) [inline app/upgrades/signetlaunch/data_signed_fps_test.go:35]: we probably do not need this assertion, as I imagine the length will be highly dependent on current launch context (devnet, testnet, mainnet)

**Review by KonradStaniec — APPROVED** [review]
Codewise LGTM!

I would say next step would be to describe in `README.md` how those strings containing data should be updated i.e:
- we are launching `devnet` how to prepare those strings and configure `app.go` so the upgrade will happen.

File: app/upgrades/signetlaunch/data_signed_fps_test.go
Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/data_signed_fps_test.go:22]: good idea, I will update

Comment (RafilxTenfen) [reply app/upgrades/signetlaunch/data_signed_fps_test.go:22]: done at https://github.com/babylonlabs-io/babylon/pull/28/commits/c503f3bf3ce59c673203f5b76b512c01f29397a4

File: app/upgrades/signetlaunch/README.md
Comment (RafilxTenfen) [inline app/upgrades/signetlaunch/README.md:5]: There is one example of shell script running this upgrade with load of BTC Headers at https://github.com/babylonlabs-io/babylon-deployment/blob/main/deployments/local/bbn-start-and-upgrade-signet-launch.sh

Did not linked in the README.md because the babylon-deployment repo is private

File: app/upgrades/signetlaunch/README.md
Comment (RafilxTenfen) [inline app/upgrades/signetlaunch/README.md:44]: https://babylonlabs.atlassian.net/wiki/x/ZoANAQ
Doc that FPs will need to follow, to create the signed MsgCreateFinalityProvider

File: app/upgrades/signetlaunch/README.md
Comment (SebastianElvis) [inline app/upgrades/signetlaunch/README.md:8]: "data point" does not seem to mean stuff in our scenario (btc headers, fps, ...)?

**Review by SebastianElvis — APPROVED** [review]

**Pull request closed without merging**
